components/resource_metering/src/reporter.rs

```diff
@@ -0,0 +1,211 @@
+        let channel = {
+            let cb = ChannelBuilder::new(self.env.clone())
+                .keepalive_time(Duration::from_secs(10))
+                .keepalive_timeout(Duration::from_secs(3));
+            self.security_mgr.connect(cb, addr)
+        };
+        self.client = Some(ResourceUsageAgentClient::new(channel));
+        if self.cpu_records_collector.is_none() {
+            self.cpu_records_collector = Some(register_collector(Box::new(
+                CpuRecordsCollector::new(self.scheduler.clone()),
+            )));
+        }
+    }
+}
+
+impl Runnable for ResourceMeteringReporter {
+    type Task = Task;
+
+    fn run(&mut self, task: Self::Task) {
+        match task {
+            Task::ConfigChange(new_config) => {
+                if !new_config.should_report() {
+                    self.client.take();
+                    self.cpu_records_collector.take();
+                } else if new_config.agent_address != self.config.agent_address
+                    || new_config.enabled != self.config.enabled
+                {
+                    self.init_reporter(&new_config.agent_address);
+                }
+
+                self.config = new_config;
+            }
+            Task::CpuRecords(records) => {
+                let timestamp_secs = records.begin_unix_time_secs;
+
+                for (tag, ms) in &records.records {
+                    let tag = &tag.infos.extra_attachment;
+                    if tag.is_empty() {
+                        continue;
+                    }
+
+                    let ms = *ms as u32;
+                    match self.records.get_mut(tag) {
+                        Some((ts, cpu_time, total)) => {
+                            if *ts.last().unwrap() == timestamp_secs {
+                                *cpu_time.last_mut().unwrap() += ms;
+                            } else {
+                                ts.push(timestamp_secs);
+                                cpu_time.push(ms);
+                            }
+                            *total += ms;
+                        }
+                        None => {
+                            self.records
+                                .insert(tag.clone(), (vec![timestamp_secs], vec![ms], ms));
+                        }
+                    }
+                }
+
+                if self.records.len() > self.config.max_resource_groups {
+                    self.find_top_k.clear();
+                    for (_, _, total) in self.records.values() {
+                        self.find_top_k.push(*total);
+                    }
+                    pdqselect::select_by(
+                        &mut self.find_top_k,
+                        self.config.max_resource_groups,
+                        |a, b| b.cmp(a),
+                    );
+                    let kth = self.find_top_k[self.config.max_resource_groups];
+                    self.records.retain(|_, (_, _, total)| *total >= kth);
```

Comment:
Can we have a benchmark for this part?

Reply:
It will cost under a microsecond and normally 2x faster than sort_unstable.